### PR TITLE
Added server returns error for PUT on file that already exists

### LINKED ISSUE #24: Server should send an error packet if the client wants to PUT a file that already exists
This could probably be accomplished in `server::Handler::put` by replacing the `create` and `truncate` options with [`OptionOptions::create_new`](https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.create_new).

### PR DESCRIPTION
**Pull request by f3rn0s**
When I was writing the test for this new functionality, I noticed that the client, on a .put(), would be moved into that put() and thus couldn't be used anymore. Maybe we need another issue to fix this?﻿

Closes #24

### CODE REVIEW

**Comment by f3rn0s** [comment]
Alright, it seems the test I wrote fails at random. I tried to diagnose the issue but I'm struggling. Here was an example of it failing after seemingly no differences at runtime:

```sh
#Command I ran
for i in {0..1000}; do ./put-3f0b54c910f16bc7; sleep 5; done
```

```text

running 2 tests
test test_put_when_already_exists ... ok
test test_put ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 2 tests
test test_put_when_already_exists ... ok
test test_put ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 2 tests
test test_put_when_already_exists ... ok
test test_put ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 2 tests
test test_put_when_already_exists ... ok
test test_put ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 2 tests
test test_put_when_already_exists ... ok
test test_put ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 2 tests
test test_put_when_already_exists ... ok
test test_put ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out


running 2 tests
test test_put ... **FAILED**
```

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Hi @f3rn0s! Thanks for the great work here.

I've left a couple of comments inline.

~~Once those are fixed, however, there's still an issue that I'm trying to debug.~~

~~The regular `test_put` test will fail when the test tries to open the file that was written because it thinks the directory does not exist. Something is removing the file too soon when the test suite runs with multiple threads. You can see it pass consistently if you run `cargo test -- --test-threads=1`.~~

~~I don't think solving that issue is within scope for this PR though, so I'd be fine if you added [`serial_test`](https://crates.io/crates/serial_test) as a dependency and put `#[serial]` on both of the `put` tests. I will file an issue to look into the multithreaded issue after the fact.~~

So, in summary:

1. Choose a different port for your test
2. Update the test to have the server handle at least 2 requests
3. ~~Add a dev-dependency on `serial_test` and make the `test_put*` tests `#[serial]`.~~

**Comment by connorkuehl** [comment]
> When I was writing the test for this new functionality, I noticed that the client, on a .put(), would be moved into that put() and thus couldn't be used anymore. Maybe we need another issue to fix this?﻿
> 

TFTP doesn't re-use connections, so I made it this way by design. However, I think your comment is a compelling reason to make the client::Builder implement Clone though. Then people can make as many clones of the builder as they want and create a Client from the clones.

**Comment by connorkuehl** [comment]
Hey @f3rn0s! My friend @squidboylan resolved the issue with the flakey test weirdness in #30, so all that's needed here is for you to rebase and address the two comments above and then this looks good to me.

**Comment by f3rn0s** [comment]
I re-based the code, as per your suggestion, off the main branch and it seems to have resolved the issue.

**Comment by connorkuehl** [comment]
It looks like there are still merge conflicts. Could you try rebasing again to fix them?

```console
$ git fetch https://github.com/connorkuehl/tftp
$ git rebase FETCH_HEAD
```

**Comment by f3rn0s** [comment]
Hi, I have resolved the conflicts.

**Comment by connorkuehl** [comment]
It looks like you solved the conflicts by merging with my branch, not rebasing on top of it. As a result, you have several commits in this pull request that aren't yours and that have already been merged into my repository. GitHub can't merge the branch like this.

I cherry-picked your commit and merged it in 3d9b6c8fdb057c6e97b1de2733119f32970c743b.

Thank you for fixing this issue! 🙂
